src/EditorFeatures/CSharp/Wrapping/SeparatedSyntaxList/CSharpArgumentWrapper.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.Editor.Wrapping.SeparatedSyntaxList
+{
+    internal partial class CSharpArgumentWrapper
+        : AbstractCSharpSeparatedSyntaxListWrapper<BaseArgumentListSyntax, ArgumentSyntax>
+    {
+        protected override string ListName => FeaturesResources.argument_list;
+        protected override string ItemNamePlural => FeaturesResources.arguments;
+        protected override string ItemNameSingular => FeaturesResources.argument;
+
+        protected override SeparatedSyntaxList<ArgumentSyntax> GetListItems(BaseArgumentListSyntax listSyntax)
+            => listSyntax.Arguments;
+
+        protected override BaseArgumentListSyntax GetApplicableList(SyntaxNode node)
+        {
+            switch (node)
+            {
+                case InvocationExpressionSyntax invocationExpression: return invocationExpression.ArgumentList;
+                case ElementAccessExpressionSyntax elementAccessExpression: return elementAccessExpression.ArgumentList;
+                case ObjectCreationExpressionSyntax objectCreationExpression: return objectCreationExpression.ArgumentList;
+                case ConstructorInitializerSyntax constructorInitializer: return constructorInitializer.ArgumentList;
+            }
+
+            return null;
+        }
+
+        protected override bool PositionIsApplicable(
+            SyntaxNode root, int position,
+            SyntaxNode declaration, BaseArgumentListSyntax listSyntax)
+        {
+            var startToken = listSyntax.GetFirstToken();
+
+            if (declaration is InvocationExpressionSyntax ||
+                declaration is ElementAccessExpressionSyntax)
+            {
+                // If we have something like  Foo(...)  or  this.Foo(...)  allow anywhere in the Foo(...)
+                // section.
```

Comment:
Yup.  This is definitely something i thought a lot about.  Here's my mental design for how the feature works:

1. there is only one 'wrapping' refactoring feature.
2. that wrapping feature is built up out of providers.  each provider can handle certain types of nodes.
3. the core feature walks upwards looking for the first provider that can supply fixes for a node up the parent-chain.
4. when we get any results, we show them and we're done.

This means:

1. you only ever get one set of wrapping options offered.
2. the wrapping options should be for the item that is the most immediately enclosing for your caret.

With cases like 'invocation wrapping' and 'argument wrapping', this means you will both get different results depending on where your caret may be and depending on the construct.  For example:

```
a.Foo($$1, 2, 3) //<-- argument wrapping offered
a.$$Foo(1, 2, 3) //<-- argument wrapping offered
a.Foo($$1, 2, 3).Bar(1, 2, 3) //<-- argument wrapping offered
a.$$Foo(1, 2, 3).Bar(1, 2, 3) //<-- fluent sequence wrapping
```

I think this will actually work out pretty well in practice.  There should always be a location where you can place the caret that will allow you to wrap something sensible.  However, in cases of non-conflicts (i.e. case #2) we can expand the region to still work well without forcing a narrow location for the caret.